packages/plugins/typed-variable-modal/src/Modal.js

```diff
@@ -0,0 +1,368 @@
+   * @private
+   */
+  handleBackwardTab_(e) {
+    if (document.activeElement === this.firstFocusableEl_) {
+      e.preventDefault();
+      this.lastFocusableEl_.focus();
+    }
+  }
+
+  /**
+   * Handle when the user goes to the next focusable element.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleForwardTab_(e) {
+    if (document.activeElement === this.lastFocusableEl_) {
+      e.preventDefault();
+      this.firstFocusableEl_.focus();
+    }
+  }
+
+  /**
+   * Handles keydown event for a Blockly modal. Handles forward tab, backward
+   * tab, and escape button.
+   * @param {KeyboardEvent} e The keydown event.
+   * @private
+   */
+  handleKeyDown_(e) {
+    if (e.keyCode === Blockly.utils.KeyCodes.TAB) {
+      // If there are no elements or there is one element don't wrap.
+      if (!this.firstFocusableEl_ ||
+          this.firstFocusableEl_ === this.lastFocusableEl_) {
+        e.preventDefault();
+        e.stopPropagation();
+        return;
+      }
+      if (e.shiftKey) {
+        this.handleBackwardTab_(e);
+      } else {
+        this.handleForwardTab_(e);
+      }
+    } else if (e.keyCode === Blockly.utils.KeyCodes.ESC) {
+      this.hide();
+    }
+    e.stopPropagation();
+  }
+
+  /**
+   * Helper method for adding an event.
+   * @param {!Element} node Node upon which to listen.
+   * @param {string} name Event name to listen to (e.g. 'mousedown').
+   * @param {Object} thisObject The value of 'this' in the function.
+   * @param {!Function} func Function to call when event is triggered.
+   * @protected
+   */
+  addEvent_(node, name, thisObject, func) {
+    const event = Blockly.bindEventWithChecks_(node, name, thisObject, func);
+    this.boundEvents_.push(event);
+  }
+
+  /**
+   * Create all the dom elements for the modal.
+   */
+  render() {
+    /*
+     * Creates the Modal. The generated modal looks like:
+     * <div class="blockly-modal-container" role="dialog">
+     *   <header class="blockly-moddal-header">
+     *     <h2 class="blockly-modal-header-title">Modal Name</h2>
+     *     <button class="blockly-modal-btn">X</button>
+     *   </header>
+     *   <div class="blockly-modal-content">
+     *   </div>
+     *   <div class="blockly-modal-footer">
+     *     <button class="blockly-modal-btn blockly-modal-btn-primary">OK</button>
+     *     <button class="blockly-modal-btn">Cancel</button>
+     *   </div>
+     * </div>
+     */
+
+    // Create Container
+    this.htmlDiv_ = document.createElement('div');
+    Blockly.utils.dom.addClass(this.htmlDiv_, 'blockly-modal-container');
+    this.htmlDiv_.setAttribute('role', 'dialog');
+    // End creating the container
+
+    // Create the header
+    const modalHeader = document.createElement('header');
+    Blockly.utils.dom.addClass(modalHeader, 'blockly-modal-header');
+
+    this.renderHeader_(modalHeader);
+
+    const exitButton = document.createElement('button');
```

Comment:
It's a good idea to place the exit button at the bottom of the DOM, otherwise its always going to be the first focusable element (which isnt a great user experience for a user using the keyboard)

Reply:
Ok, I removed the X in the top right corner and changed the cancel button to be in the modal.

Reply:
Curious why you decided to remove it?

Reply:
Talked about offline a bit. In the end I put the 'x' button back and kept it in the header. When I show the modal I skip over the 'x' button and focus the next item in the list if one exists. The reason I didn't position absolutely is because I didn't have all the positioning info I needed. Also, if I positioned absolutely and someone had a very long title then the button would overlap with it.